### ISSUE TITLE
CmyTime Rebuild: JAX-based C. elegans Learning Simulation

### ISSUE THREAD

**Issue by daniyal-rahman** [daniyal-rahman/CmyTime#2]
# CmyTime Rebuild Plan: JAX-based C. elegans Learning Simulation

## Overview

Rebuild the project from scratch using JAX to create a C. elegans neural network simulation with STDP learning, enabling the worm to learn food-seeking and poison-avoidance behaviors.

## Git Branch Strategy

1. Create new branch `fresh-start` from current state
2. Archive current work: rename `main` to `archive-openworm-integration`
3. Make `fresh-start` the new `main`
4. Delete from new main:
   - `Sibernetic/` (C++ physics - replacing with Python)
   - `org.geppetto/` (Java viz - not needed)
   - `venv/` (will create fresh)
   - `muscle_activations.csv` (generated file)
   - `README.md` (empty)
   - `examples` (unclear file)

5. Keep entire `c302/` directory as read-only reference:
   - Do not modify any c302 files
   - Use for connectome data extraction only
   - Allows potential upstream updates if needed

6. Keep documentation:
   - `CLAUDE.md` (will update)
   - `PROJECT_GOALS.md` (will update)
   - `main.md`, `curr.md` (reference)

## Fresh Virtual Environment

Create new venv with minimal dependencies:
```bash
python -m venv venv
source venv/bin/activate
pip install jax jaxlib numpy matplotlib pygame pytest
```

Core dependencies:
- `jax` + `jaxlib` - Neural simulation, autodiff, JIT
- `numpy` - Data handling, connectome loading
- `matplotlib` - Static plots, analysis figures
- `pygame` - Real-time interactive visualization
- `pytest` - Testing

## New Project Structure

```
src/
├── celegans/
│   ├── __init__.py
│   │
│   ├── neural/                 # Core neural simulation
│   │   ├── __init__.py
│   │   ├── neuron.py          # LIF neuron dynamics
│   │   ├── synapse.py         # Chemical synapses + STDP
│   │   ├── gap_junction.py    # Electrical coupling
│   │   ├── network.py         # Full network integration
│   │   └── connectome.py      # Load from c302 data
│   │
│   ├── body/                   # 2D articulated physics
│   │   ├── __init__.py
│   │   ├── segments.py        # Body segment representation
│   │   ├── kinematics.py      # Position/angle computations
│   │   └── locomotion.py      # Thrust from undulation
│   │
│   ├── motor/                  # Motor control
│   │   ├── __init__.py
│   │   ├── muscles.py         # 96 muscle groups
│   │   └── motor_map.py       # Motor neurons → muscles → curvature
│   │
│   ├── sensory/                # Sensory system
│   │   ├── __init__.py
│   │   ├── chemosensory.py    # Gradient sensing
│   │   └── sensory_map.py     # Environment → sensory neurons
│   │
│   ├── environment/            # 2D arena
│   │   ├── __init__.py
│   │   ├── arena.py           # Boundaries, collision
│   │   ├── stimuli.py         # Food/poison sources
│   │   └── gradient.py        # Concentration fields
│   │
│   ├── learning/               # Learning rules
│   │   ├── __init__.py
│   │   ├── stdp.py            # Basic STDP
│   │   ├── reward_stdp.py     # Reward-modulated STDP
│   │   └── eligibility.py     # Eligibility traces
│   │
│   ├── simulation/             # Main loop
│   │   ├── __init__.py
│   │   ├── state.py           # SimulationState container
│   │   ├── step.py            # Single timestep
│   │   └── runner.py          # Multi-step execution
│   │
│   └── utils/
│       ├── __init__.py
│       ├── config.py          # All hyperparameters
│       ├── plotting.py        # Matplotlib static analysis plots
│       ├── live_viz.py        # Pygame real-time visualization
│       └── checkpointing.py   # Save/load

├── scripts/
│   ├── extract_connectome.py  # One-time c302 data extraction
│   ├── run_simulation.py      # Main entry point
│   └── analyze_learning.py    # Post-hoc analysis

├── tests/
│   ├── test_neural.py
│   ├── test_body.py
│   ├── test_motor.py
│   └── test_integration.py

└── data/
    └── connectome.npz         # Extracted connectome (generated)
```

### Neural State
```python
class NeuronState(NamedTuple):
    V: jnp.ndarray              # (302,) membrane potentials
    spike: jnp.ndarray          # (302,) binary spike flags
    refractory: jnp.ndarray     # (302,) refractory counters

class SynapseState(NamedTuple):
    W: jnp.ndarray              # (N_chem,) chemical synapse weights
    pre_trace: jnp.ndarray      # (302,) STDP pre traces
    post_trace: jnp.ndarray     # (302,) STDP post traces

class GapJunctionState(NamedTuple):
    G: jnp.ndarray              # (N_gj,) conductances (fixed)
```

### Body State
```python
class BodyState(NamedTuple):
    theta: jnp.ndarray          # (24,) segment angles
    theta_dot: jnp.ndarray      # (24,) angular velocities
    head_pos: jnp.ndarray       # (2,) [x, y]
    head_angle: float           # orientation
```

### Connectome (Static, loaded once)
```python
class Connectome(NamedTuple):
    neuron_names: tuple         # 302 names
    neuron_types: jnp.ndarray   # 0=sensory, 1=inter, 2=motor
    chem_pre: jnp.ndarray       # chemical synapse pre indices
    chem_post: jnp.ndarray      # chemical synapse post indices
    chem_type: jnp.ndarray      # 0=excitatory, 1=inhibitory
    gj_neuron1: jnp.ndarray     # gap junction indices
    gj_neuron2: jnp.ndarray
```

### 1. LIF Neuron Update
```python
def lif_step(V, spike, refractory, I_total, params, dt):
    dV = ((V_rest - V) / tau_m + I_total / C) * dt
    V_new = where(refractory > 0, V_reset, V + dV)
    spike_new = (V_new >= V_thresh)
    V_new = where(spike_new, V_reset, V_new)
    return V_new, spike_new, updated_refractory
```

### 2. Gap Junction Current (bidirectional)
```python
def gap_junction_current(V, G, connectome):
    V1, V2 = V[connectome.gj_neuron1], V[connectome.gj_neuron2]
    I_flow = G * (V1 - V2)
    # Scatter to both neurons (opposite signs)
    I = zeros(302)
    I = I.at[gj_neuron1].add(-I_flow)
    I = I.at[gj_neuron2].add(+I_flow)
    return I
```

### 3. Reward-Modulated STDP
```python
def rstdp_update(W, pre_trace, post_trace, pre_spike, post_spike,
                 reward, eligibility, connectome, params, dt):
    # Decay and update traces
    pre_trace = pre_trace * exp(-dt/tau_pre) + pre_spike
    post_trace = post_trace * exp(-dt/tau_post) + post_spike

    # STDP signal -> eligibility
    stdp = A_plus * pre_trace[pre_idx] * post_spike[post_idx] \
         - A_minus * post_trace[post_idx] * pre_spike[pre_idx]
    eligibility = eligibility * exp(-dt/tau_e) + stdp

    # Weight update gated by reward
    dW = learning_rate * reward * eligibility
    W_new = clip(W + dW, W_min, W_max)
    return W_new, pre_trace, post_trace, eligibility
```

### 4. Motor → Muscle → Curvature
```python
def motor_to_curvature(motor_activity, motor_map):
    # Motor neurons (VA, VB, DA, DB, DD, VD) → 96 muscles
    muscle_act = map_motors_to_muscles(motor_activity, motor_map)

    # Muscles → curvature (dorsal - ventral difference)
    curvature = zeros(24)
    for row in range(24):
        dorsal = muscle_act[MDR[row]] + muscle_act[MDL[row]]
        ventral = muscle_act[MVR[row]] + muscle_act[MVL[row]]
        curvature[row] = gain * (dorsal - ventral)
    return curvature
```

### 5. Body Kinematics
```python
def body_step(theta, theta_dot, head_pos, head_angle, target_curvature, params, dt):
    # Spring-damper dynamics toward target curvature
    theta_ddot = stiffness * (target_curvature - theta) - damping * theta_dot
    theta_dot_new = theta_dot + theta_ddot * dt
    theta_new = theta + theta_dot_new * dt

    # Thrust from traveling wave
    thrust = compute_thrust(theta_new, theta_dot_new)
    head_pos_new = head_pos + thrust * dt
    head_angle_new = head_angle + theta_dot_new[0] * dt
    return theta_new, theta_dot_new, head_pos_new, head_angle_new
```

### 6. Sensory Input
```python
def sensory_input(head_pos, head_angle, environment, sensory_map):
    # Food gradient at head
    food_grad = grad(food_concentration)(head_pos)
    poison_grad = grad(poison_concentration)(head_pos)

    # Project to left/right sensory pairs
    local_grad = rotate_to_head_frame(food_grad, head_angle)
    awc_left = local_grad[0] + local_grad[1]   # forward + lateral
    awc_right = local_grad[0] - local_grad[1]  # forward - lateral

    # Map to sensory neuron indices
    I_sensory = zeros(302)
    I_sensory = I_sensory.at[sensory_map.AWCL].set(awc_left)
    I_sensory = I_sensory.at[sensory_map.AWCR].set(awc_right)
    # ... ASH for poison, etc.
    return I_sensory
```

### 7. Reward Signal
```python
def compute_reward(head_pos, prev_head_pos, environment, params):
    food_now = food_concentration(head_pos)
    food_prev = food_concentration(prev_head_pos)
    poison_now = poison_concentration(head_pos)
    poison_prev = poison_concentration(prev_head_pos)

    # Reward for approaching food, penalty for approaching poison
    reward = food_delta_weight * (food_now - food_prev) \
           - poison_delta_weight * (poison_now - poison_prev)

    # Bonus/penalty for reaching targets
    if distance_to_nearest_food < threshold:
        reward += food_bonus
    if distance_to_nearest_poison < threshold:
        reward -= poison_penalty

    return reward
```

### Phase 1: Connectome Extraction (Day 1)
1. Write `scripts/extract_connectome.py`
   - Load from `c302/c302/data/herm_full_edgelist.csv`
   - Parse chemical synapses and gap junctions
   - Classify neurons (sensory/inter/motor) by naming convention
   - Save to `data/connectome.npz`

### Phase 2: Neural Simulator (Days 2-4)
1. `neural/neuron.py` - LIF dynamics
2. `neural/synapse.py` - Chemical synapse current
3. `neural/gap_junction.py` - Electrical coupling
4. `neural/connectome.py` - Load connectome.npz
5. `neural/network.py` - Integrate all components
6. `tests/test_neural.py` - Verify dynamics

### Phase 3: Body Physics (Days 5-6)
1. `body/segments.py` - 24-segment articulated body
2. `body/kinematics.py` - Forward kinematics
3. `body/locomotion.py` - Thrust computation
4. `tests/test_body.py` - Verify undulation produces movement

### Phase 4: Motor Mapping (Day 7)
1. `motor/muscles.py` - 96 muscle representation
2. `motor/motor_map.py` - Motor neuron → muscle → curvature
3. Integration test: Pattern of motor activity → locomotion

### Phase 5: Environment & Sensory (Days 8-9)
1. `environment/arena.py` - 2D bounded arena
2. `environment/stimuli.py` - Food/poison sources
3. `environment/gradient.py` - Concentration fields
4. `sensory/chemosensory.py` - Gradient sensing
5. `sensory/sensory_map.py` - Map to sensory neurons

### Phase 6: Learning (Days 10-11)
1. `learning/stdp.py` - Basic STDP
2. `learning/eligibility.py` - Eligibility traces
3. `learning/reward_stdp.py` - R-STDP integration
4. Test: Weights change appropriately with reward

### Phase 7: Integration (Days 12-14)
1. `simulation/state.py` - Complete state container
2. `simulation/step.py` - Full closed-loop step
3. `simulation/runner.py` - jax.lax.scan for efficiency
4. `utils/visualization.py` - Plotting utilities
5. End-to-end test: Learning to find food

### Phase 8: Tuning & Validation (Days 15+)
1. Parameter tuning for stable learning
2. Validate emergent behaviors
3. Documentation and cleanup

## Key Parameters (Tunable)

```python
# Neuron
V_rest = -65.0 mV
V_thresh = -50.0 mV
tau_m = 20.0 ms

# STDP
tau_pre = tau_post = 20.0 ms
A_plus = 0.01
A_minus = 0.012  # Slightly larger for stability

# R-STDP
tau_eligibility = 1000.0 ms
learning_rate = 0.001

# Body
n_segments = 24
stiffness = 100.0
damping = 10.0

# Reward
food_delta_weight = 10.0
poison_delta_weight = 15.0
food_bonus = 100.0
poison_penalty = 100.0
```

## Critical Source Files (Read Before Implementation)

1. `c302/c302/data/herm_full_edgelist.csv` - Actual connectivity data
2. `c302/c302/ConnectomeReader.py` - PREFERRED_NEURON_NAMES (302 neurons)
3. `c302/c302/SpreadsheetDataReader.py` - Data parsing patterns
4. `c302/c302/__init__.py` - Motor neuron positions (VB_soma_pos, DB_soma_pos)
5. `Sibernetic/main_sim.py` - Muscle quadrant layout reference

## Notes

- **Neuron classification**: Using c302 naming conventions for now (VA*=motor, etc.). TODO: Make explicit mapping file later for robustness.
- **Gap junctions are critical**: ~600 electrical synapses provide fast coupling between neurons - must include these.
- **Reward signal**: Using change in concentration (approaching food = positive) rather than just proximity, to give directional learning signal.
